// @kyve/core/dist/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var chalk_1 = __importDefault(require("chalk"));
var fecha_1 = __importDefault(require("fecha"));
var Log = /** @class */ (function () {
    function Log(name) {
        this.name = name;
    }
    Log.prototype.error = function (message) {
        console.log(this.formatMessage(message, "error"));
    };
    Log.prototype.warn = function (message) {
        console.log(this.formatMessage(message, "warn"));
    };
    Log.prototype.info = function (message) {
        console.log(this.formatMessage(message, "info"));
    };
    Log.prototype.debug = function (message) {
        console.log(this.formatMessage(message, "debug"));
    };
    Log.prototype.formatMessage = function (message, level) {
        var formattedName = this.getFormattedName();
        var formattedLevel = this.getFormattedLevel(level);
        var formattedTiemstamp = this.getFormattedTimestamp();
        return formattedTiemstamp + " " + formattedLevel + "[" + formattedName + "]:  " + message;
    };
    Log.prototype.getFormattedName = function () {
        return chalk_1["default"].bold.whiteBright(this.name);
    };
    Log.prototype.getFormattedLevel = function (level) {
        switch (level) {
            case "error":
                return chalk_1["default"].bold.redBright(level);
            case "warn":
                return chalk_1["default"].bold.yellowBright(level);
            case "info":
                return chalk_1["default"].bold.greenBright(level);
            case "debug":
                return chalk_1["default"].bold.cyanBright(level);
        }
    };
    Log.prototype.getFormattedTimestamp = function () {
        var timestamp = fecha_1["default"].format(new Date(), "HH:mm:ss");
        return chalk_1["default"].grey(timestamp);
    };
    return Log;
}());
exports["default"] = Log;
